further improved the observability strategy document (#713)
kyma-project/community: concepts/observability-strategy/strategy.md

# From Observability to Telemetry - a strategy shift in Kymas Observability feature

## Motivation

> The Kyma project evolved over time, from an extensibility framework having batteries-included, to a runtime which focusses on providing kubernetes building blocks with enterprise-grade quality. Hereby, the Kyma observability feature was not adapting yet, still providing an out-of-the box in-cluster stack which gets you started immediately, but does not qualify for the enterprise-grade qualities.

Reviewer: ```suggestion
The Kyma project evolved over time, from an extensibility framework with "batteries included", to a runtime that focuses on providing Kubernetes building blocks with enterprise-grade quality. Up until now, the Kyma Observability feature was still providing an out-of-the box in-cluster stack that gets you started immediately, but which does not qualify for the enterprise-grade qualities. Now, the the Kyma Observability feature is evolving, too.
```

> Observability backends are available in various ways from different vendors as managed services. While users can pick and choose and can leverage benefits like cross-cluster scenarios, the integration aspect into that external managed services is often complex and heterogenous. Kyma should not invest into another opinionated solution for an in-cluster backend with enterprise-grade qualities, it should instead focus on simplifying the instrumentation and collection of the telemetry data and the integration of the data into external solutions. Hereby, it will leverage as much as possible the OpenTelemetry standards to provide integration options in a vendor-neutral approach.

Reviewer: ```suggestion
Various observability backends are available as managed services from different vendors. While users can pick and choose and can leverage benefits like cross-cluster scenarios, the integration with such externally managed services is often complex and heterogenous. Kyma should not invest in another opinionated solution for an in-cluster backend with enterprise-grade qualities. Instead, it should focus on simplifying the instrumentation and collection of the telemetry data, and the integration of the data into external solutions. Hereby, Kyma will as much as possible leverage the OpenTelemetry standards to provide integration options in a vendor-neutral approach.
```

> With that Kyma will shift the focus away from providing a full Observability solution, instead it will simplify the instrumentation and integration of telemetry data into external systems.

Reviewer: ```suggestion
With that, Kyma will shift the focus away from providing a full Observability solution; instead it will simplify the instrumentation and integration of telemetry data into external systems.
```

## Motivation

The Kyma project evolved over time, from an extensibility framework with "batteries included", to a runtime that focuses on providing Kubernetes building blocks with enterprise-grade quality. Up until now, the Kyma Observability feature is providing an out-of-the box in-cluster stack that gets you started immediately, but which does not qualify for the enterprise-grade qualities. Various observability backends are already available as managed services from different vendors fulfilling these criterias. The possibility for integration into these offerings will not only bring provide the missing value but also Openness - it is up to the users scenarios which observability system will fit best to the actual requirements.
When Kyma supports a simple way of instrumenting, collecting and integration of the workloads telemetry data into available observability systems, leveraging the [OpenTelemetry](https://opentelemetry.io/) standards to provide integration options in a vendor-neutral approach, the Kyma Observability feature is ready for it's evolution too.

> With that, Kyma will shift the focus away from providing a full Observability solution; instead it will simplify the instrumentation and integration of telemetry data into external systems and with that will provide the support which is needed by keeping biggest flexibility.

Reviewer: ```suggestion
With that, Kyma will shift the focus away from providing a full Observability solution; instead it will simplify the instrumentation and integration of telemetry data into external systems. Thus, Kyma will keep providing support while at the same time, providing the biggest flexibility.
```

### Drawbacks

> At a first glance, the current solution provides a feature-rich end-to-end setup. However, at a second glance, users notice major drawbacks which are not fitting to enterprise-grade software and usually need additional stacks.

Reviewer: ```suggestion
At a first glance, the current solution provides a feature-rich end-to-end setup. However, at a second glance, users notice major drawbacks that aren't a good fit for enterprise-grade software and usually need additional stacks.
```

- Storage backends are non-scalable, so they can be used only in limited scenarios. Users cannot upgrade the backend into a scalable setup, nor integrate with other solutions.

## Requirements

Integration (and with that, moving the focus away from providing backends) is the key to open up the stack for a broad range of use cases. Users can bring their own backends if they already use a commercial offering or run their own infrastructure. To name just a few advantages, the data can be stored outside the cluster in a managed offering, shared with the data of multiple clusters, and kept away from any tampering or deletion attempts by hackers.

Providing ready-to-use in-cluster backends is necessarily opinionated, does not cover all usage scenarios, and does not fit into Kyma's goal of providing the Kubernetes building blocks to integrate into the the SAP ecosystem. Also, the licensing issues (particularly with Grafana and Loki, or Elasticsearch as an alternative backend technology) show that an opinionated stack is problematic. It's better to handle opinionation by integrating with actual managed services.

> Instrumentation and integration of telemetry data into external systems can be complex and is heterogenous. Providing instrumentation conventions so that provided collectors can automatically pick-up the data and make the integration problem a central aspect, can simplify that tasks a lot.

Reviewer: ```suggestion
Instrumentation and integration of telemetry data into external systems can be complex and is heterogenous. To simplify those tasks, Kyma will provide instrumentation conventions so that provided collectors can automatically pick up the data and make the integration problem a central aspect.
```

_Who_ will make the integration problem a central aspect? The collectors, as the current sentence implies?

## Architecture

> The strategy shift will be backed by a new layer of collectors that are not bound to any backend anymore. This layer is responsible for collecting and enriching all telemetry data, depending on the signal type. The data will be collected automatically when the best practices for instrumentation are followed.

Reviewer: ```suggestion
The strategy shift will be backed by a new layer of collectors that are not bound to any backend anymore. This layer is responsible for collecting and enriching all telemetry data, depending on the signal type. As long as the best practices for instrumentation are followed, the data is collected automatically.
```

Users can configure the collectors at runtime with different signal pipelines using basic filtering (inclusion and exclusion of signals) and outputs, so that the collectors start shipping the signals through the pipelines to the configured backends. The dynamic configuration and management of the collector is handled by a new operator, which is configured using Kubernetes API. The collectors and the new operator are bundled in a new core component called `telemetry`.

> The configuration options for the collectors via the Kubernetes API will be limited in order to guarantee enterprise-grade qualities. However, it will be possible to run your own collector setup for advanced customization options.

Reviewer: ```suggestion
To guarantee enterprise-grade qualities, the configuration options for the collectors using the Kubernetes API will be limited. However, users can run their own collector setup for advanced customization options.
```

The existing Kyma backends and UIs will be just one possible solution to integrate with. Users will still be able to install them manually with a blueprint.

![b](./assets/strategy-future.drawio.svg)

The technology stack for instrumentation and collection will be based on the OpenTelemetry project. The central data protocol will be OTLP, for trace propagation it will be w3c-tracecontext. As exception to that, the technology stack for collection logs will be based on the Fluent ecosystem and the specific protocols with the goal to adopt to OpenTelemetry at a later time.

As mentioned before, the collector setup and used protocols for the signal collection depend on the signal types: Logs are tailed from container log files, metrics usually are pulled using the Prometheus format, and traces are pushed with OTLP. With that, the pre-integration (so that typical signals are collected instantly) is different per type.

> That's why the specific concepts for the different types are different, and are discussed in more detail in the following documents which are still under development and should not be seen final yet:

Reviewer: ```suggestion
That's why the specific concepts for the different types are different, and are discussed in more detail in the following documents. Note that these documents are still under development and should not be seen as final yet:
```

- **Custom output protocols can be integrated**: Users can have custom transformation from the supported protocol, usually with additional operational effort (like running a dedicated Otel Collector sidecar or deployment in the users' Namespace). Furthermore, Kyma offers typical output plugins of the used collector technology as part of a pipeline definition; however, only with limited support.
- **Blueprints are fully integrated**: Blueprints for local backend deployments are integrated with the used pipeline mechanism, so that they work instantly.

## Conclusion

> Shifting the focus of Kymas Observability feature away from an end-to-end observability solution having major drawbacks (in-cluster, opinionated, not flexible enough, ..), into a solution which supports the user with the runtime centric tasks of instrumentation, collection and integration of telemetry data into external systems in a vendor-neutral way, which solves an actual problem of enterprise users - getting telemetry data reliable into an external system with few efforts.

Reviewer: ```suggestion
Shifting the focus of Kyma's Observability feature away from an end-to-end observability solution that has major drawbacks (in-cluster, opinionated, not flexible enough, ..), to a solution that supports the user with the runtime-centric tasks of instrumentation, collection, and integration of telemetry data into external systems in a vendor-neutral way, which solves an actual problem of enterprise users - getting telemetry data reliable into an external system with few efforts.
```

Even though this is a very very long sentence, it has no predicate. What do you want to say with this? We need to split it in 2-3 shorter sentences.

Reviewer: Has been rephrased
